### ISSUE TITLE
Use OCR to support structuring in metadata editor

### ISSUE THREAD

**Issue by oliver-stoehr** [kitodo/kitodo-production#5476]
The process of structuring in the metadata editor should be supported by a small tool to increase the level of automation.
It should be possible to select one or multiple pages containing the table of contents and perform OCR them.
The result of the OCR should be used to create structure elements and semi automatically insert the text. The workflow in the editor could be:
1. The user selects the page containing the table of contents.
2. The user clicks a new button "automatic structuring". This performs the OCR on the selected page.
3. The user can view the analysed page and click on one entry of the table of contents. Clicking on one entry will copy the recognised text.
4. A dialog will appear to select the type of the structure element to be created.
5. The user should click into the structure tree to select the position of the new structure element.
6. The new structure element is created and the selected text from step 3 is pasted into the title metadata of the element.

Requirement for this feature would be that @OCR-D is implemented in Kitodo.Production to allow a convenient way to perform the OCR.

**Comment by illipsum** [comment]
Very good idea! I strongly support this feature request.

**Comment by markusweigelt** [comment]
Currently an implementation project is underway for the integration of OCR-D and Kitodo. https://github.com/slub/ocrd_kitodo

In the context of our discussion, automated structuring was a topic, but for reasons not a main priority yet.

However, this issue shows us that there is a need for this. We will discuss this and possible include it in the integration.

**Comment by bertsky** [comment]
Additional ideas by @BartChris (originally reported to [ocrd_kitodo](https://github.com/slub/ocrd_kitodo/issues/60#issuecomment-1412475075)):

- (implementation) instead of a fully functional ALTO viewer, build a poor man's viewer which simply parses the `TextString` into a list of text lines and presents them in a text editor pane
- (feature) could also be used in a "preview" context (outside of the metadata editor)
- (feature) indicate which pages have been processed already (by looking into the filesystem, which image files there are `FULLTEXT` files for)
- (feature) when adding/removing/moving pages, change the existing `FULLTEXT` files as well (so as to keep in sync)

**Comment by BartChris** [comment]
@solth I would like to suggest the UI parts of the OCR-D integration for the development fund but i cannot assign labels. Could you mark the issue maybe as candidate for the development fund?

**Comment by solth** [comment]
> @solth I would like to suggest the UI parts of the OCR-D integration (needs more precision on what is needed) for the development fund but i cannot assign labels. Could you mark the issue maybe as candidate for the development fund?

@BartChris I updated your role in the repository - could you check if you can assign labels to issues again?

**Comment by BartChris** [comment]
@solth Great, it works now.

**Comment by aetherfaerber** [comment]
> > In the context of our discussion, automated structuring was a topic, but for reasons not a main priority yet.
> 
> To be precise, what we discussed so far is a _different scenario_: fully automatic structuring – as an external script task (backed by OCR-D) at the end of the Production workflow, when the METS has already been exported and can therefore be amended by automatic structuring tools (which naturally operate on METS/MODS directly).

Thank you for the clarification. I have made a suggestion towards such fully automatic structuring at https://github.com/kitodo/kitodo-production/issues/5573#issuecomment-1452135114 albeit coming from a different scenario. As far as I understand the implementation already present in the BAR's fork modifies the metadata **before** the images are loaded into the metadata editor (since the separator pages are removed before that).
Also, as far as I can see one is free to access the same METS file that Kitodo is going to write metadata entered in the metadata editor into through the `metadata` folder and change its contents at any point as one sees fit.
Could you explain why you think the file could/should only be amended after the export?
I would like to know how different this two scenarios are and if joining the suggestions makes any sense or if we would need a completely new proposal for “fully automatic structuring by OCR”.

**Comment by bertsky** [comment]
> Also, as far as I can see one is free to access the same METS file that Kitodo is going to write metadata entered in the metadata editor into through the `metadata` folder and change its contents at any point as one sees fit.

No, according to @Kathrin-Huber IIRC that file (and its schema) is:
- parsed back on the fly into the runtime model; the parser cannot cope with everything you can do in METS/MODS
- not a stable format but subject to change with every version as developers see fit

So even if you can work out something for a particular version, it might break with the next update.

> Could you explain why you think the file could/should only be amended after the export?

That's the only reason. An external tool cannot know our exact schema. (The additions to Kitodo you sketched in #5573 – scanning physical barcode inserts with structuring information – would require an internal structuring tool, so that can be anywhere in the workflow.)

IMO the 3 scenarios are quite different:
1. semi-automatic support for structuring (this PR) just does a preliminary OCR of single pages: whereever the OCR is placed in the workflow (as a script task)
2. fully automatic structuring via OCR results runs visual-textual analysis tools across the entire document and tries to infer the structMap: after export, for the reason given above
3. fully automatic structuring via barcode pages just decodes the insert pages, it needs quite a few UI extensions and should be tightly integrated into the structure model: should be an internal tool

**Comment by BartChris** [comment]
Partly related: https://github.com/kitodo/kitodo-production/issues/3837

**Comment by aetherfaerber** [comment]
>     2. fully automatic structuring via OCR results runs visual-textual analysis tools across the entire document and tries to infer the structMap: after export, for the reason given above

Just to make sure: your point is also that fully automatic structuring happens after Kitodos job is finished and therefore can't be considered in scope for Kitodo. It follows that a development fund proposal would not make any sense and this has to be tackled in an external project (OCR-D). Correct?
